tracing/src/main/java/com/palantir/tracing/Trace.java

```diff
@@ -28,62 +33,209 @@
  * Represents a trace as an ordered list of non-completed spans. Supports adding and removing of spans. This class is
  * not thread-safe and is intended to be used in a thread-local context.
  */
-public final class Trace {
+public abstract class Trace {
 
-    private final Deque<OpenSpan> stack;
-    private final boolean isObservable;
     private final String traceId;
 
-    private Trace(ArrayDeque<OpenSpan> stack, boolean isObservable, String traceId) {
-        checkArgument(!traceId.isEmpty(), "traceId must be non-empty");
-
-        this.stack = stack;
-        this.isObservable = isObservable;
+    private Trace(String traceId) {
+        checkArgument(!Strings.isNullOrEmpty(traceId), "traceId must be non-empty");
         this.traceId = traceId;
     }
 
-    Trace(boolean isObservable, String traceId) {
-        this(new ArrayDeque<>(), isObservable, traceId);
+    /**
+     * Opens a new span for this thread's call trace, labeled with the provided operation and parent span. Only allowed
+     * when the current trace is empty.
+     */
+    final OpenSpan startSpan(String operation, String parentSpanId, SpanType type) {
```

Comment:
Is worth adding an error-prone `@MustUse` annotation to these methods, and recommending `fastStartSpan` if you don't actually need the OpenSpan object?

Reply:
I did that on the public APIs, I can add annotations here, though the sampled tracer delegates internally and would need to suppress. Probably for the best in order to avoid internal misuse, I’ll update